polkadot/node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -1080,22 +1087,48 @@ impl Initialized {
 			}
 		};
 
+		let n_validators = env.validators().len();
+
 		gum::trace!(
 			target: LOG_TARGET,
 			?candidate_hash,
 			?session,
-			num_validators = ?env.session_info().validators.len(),
+			?n_validators,
 			"Import result ready"
 		);
 		let new_state = import_result.new_state();
 
+		let byzantine_threshold = polkadot_primitives::byzantine_threshold(n_validators);
+		// combine on-chain with off-chain disabled validators
+		let disabled_validators = {
+			let mut d: HashSet<ValidatorIndex> = HashSet::new();
+			for v in env
+				.disabled_indices()
+				.iter()
+				.cloned()
+				.chain(self.offchain_disabled_validators.iter(session))
+			{
+				// process disabled validators in the following order:
```

Comment:
> We might (and we do rn) slash different amount for backers backing invalid and validators voting valid on invalid candidates. This logic here does not seem to make a distinction between those two.

If the validators are slashed at all, they will be disabled onchain (in which case we will take the highest offenders). Here we account for the fact when we can't disable onchain (past era) and use offchain state in addition to that.

> Implementation/maintainability detail but can we somehow abstract the specific offence types away? Maybe there will be more offences or different offences later on. All we should care about is that we select the top offenders up to 1/3 no matter what they did.

It is somewhat abstracted in `offchain_disabled_validators.iter(session)`, but internally it has to know about different types of offences. We do indeed have some duplication in node and runtime side logic and it might diverge. I am not sure if we can eliminate entirely. Node side can't predict what is going to happen in the runtime wrt to slash ordering. We could emulate it, but if a simple ordering gets away with spam protection, I'd say it's good enough.

Reply:
take a look at [6f1c071](https://github.com/paritytech/polkadot-sdk/pull/2637/commits/6f1c07174acd0c889fd1de738beeba7f1dc56990)